index.d.ts

```diff
@@ -1,288 +1,600 @@
+    flatColumns: ColumnInstance<D>[]
+    headerGroups: HeaderGroup<D>[]
+    headers: ColumnInstance<D>[]
+    flatHeaders: ColumnInstance<D>[]
+    rows: Row<D>[]
+    getTableProps: (props?: object) => object
+    prepareRow: (row: Row<D>) => void
+    rowPaths: string[]
+    flatRows: Row<D>[]
+  }
+
+  export interface UseTableHeaderGroupProps<D extends object> {
+    headers: ColumnInstance<D>[]
+    getHeaderGroupProps: (props?: object) => object
+    // Added by helpers
+    totalHeaderCount: number
   }
 
-  export interface Row<D = {}> {
+  export interface UseTableColumnProps<D extends object> {
+    id: IdType<D>
+    isVisible: boolean
+    render: (type: 'Header' | string, props?: object) => ReactNode
+    getHeaderProps: (props?: object) => object
+    // added by default
+    parent: ColumnInstance<D>
+    depth: number
     index: number
+  }
+
+  export interface UseTableRowProps<D extends object> {
     cells: Cell<D>[]
-    getRowProps: (userProps?: any) => any
+    values: Record<IdType<D>, CellValue>
+    getRowProps: (props?: object) => object
+    index: number
     original: D
-    path: any[]
-    values: any[]
-    depth: number
+    path: IdType<D>[]
+    subRows: Row<D>[]
+    state: object
   }
 
-  export interface UseExpandedRow<D = {}> {
-    subRows?: D[]
-    groupByID?: string | number
-    toggleExpanded?: () => any
-    isExpanded?: boolean
-    isAggregated?: boolean
+  export interface UseTableCellProps<D extends object> {
+    column: ColumnInstance<D>
+    row: Row<D>
+    value: CellValue
+    getCellProps: (props?: object) => object
+    render: (type: 'Cell' | string, userProps?: object) => ReactNode
   }
 
-  export type AccessorFn<D> = (
-    originalRow: D,
-    index: number,
-    sub: {
-      subRows: [D]
-      depth: number
-      data: [D]
-    }
-  ) => unknown
-
-  export interface HeaderColumn<D> {
-    /**
-     * This string/function is used to build the data model for your column.
-     */
-    accessor: IdType<D> | AccessorFn<D>
-    Header?: ReactNode | ((props: TableInstance<D>) => ReactNode)
-    Filter?: ReactNode | ((props: TableInstance<D>) => ReactNode)
-    Cell?: ReactNode | ((cell: Cell<D>) => ReactNode)
-    /**
-     * This is the unique ID for the column. It is used by reference in things like sorting, grouping, filtering etc.
-     */
+  type HeaderProps<D extends object> = TableInstance<D> & {
+    column: ColumnInstance<D>
+  } & Record<string, any>
+  type CellProps<D extends object> = TableInstance<D> & {
+    column: ColumnInstance<D>
+    row: Row<D>
+    cell: Cell<D>
+  } & Record<string, any>
+
+  // NOTE: At least one of (id | accessor | Header as string) required
+  export interface Accessor<D extends object> {
+    accessor?:
+      | IdType<D>
+      | ((
+          originalRow: D,
+          index: number,
+          sub: {
+            subRows: D[]
+            depth: number
+            data: D[]
+          }
+        ) => string)
```

Comment:
Docs says:

> accessor: String | Function

not sure what's correct tho.

Reply:
`accessor` is a string or a function that returns a string.

Reply:
Misread the definition, cool 👍